requests-for-comments/2023-11-22-scheduled-flag-changes.md

```diff
@@ -0,0 +1,84 @@
+# Request for comments: Scheduled Feature Flag changes
+
+## Problem
+
+Currently, users can make updates to feature flags only manually and in real-time. This is not ideal, as it potentially leads to overlooked or forgotten changes, particularly when they fall outside normal working hours. This can lead to poorly timed releases, negatively impacting end users' engagement and feedback.
+
+## Solution
+
+Scheduling feature flag changes allows users to automatically change flag properties at a future point in time. For example, a user can schedule an initial release of the flag for internal testing, then enable it to the beta users group 2 days later, and finally roll it out to 100% of the users in 2 weeks.
+
+## Scope
+
+This feature should allow the following scheduled modifications to a feature flag:
+
+-   **Enable/disable the flag** (field: `is_active`)
+-   **Add rollout conditions to the flag** (field: `filters`)
+
+We want to be able to _generalize_ this implementation to other models such as `Survey` or `Experiment`.
+
+## Implementation
+
+To schedule future changes, we can have a scheduling API, specifying the model the operate on, the id, the changes to be executed, and the schedule time.
+
+```
+POST /api/projects/:project_id/schedule
+{
+    "model": "FeatureFlag",
+    "id": 57,
+    "changes": [ ... ]
+    "scheduled_at": <timestamp>
+}
+```
+
+A change to the record can be expressed as a JSON, specifying the field to be modified and the new value to be set:
+
+**Example 1: Enable the flag**
+
+```
+{
+    "field": "is_active",
+    "value": "false"
+}
+```
+
+**Example 2: Add a rollout condition to the flag**
+
+```
+{
+    "field": “filter”,
+    "value": {
+        "variant": null,
+        "properties": [{
+            "key": "$geoip_country_name",
+            "type": "person",
+            "value": ["Australia"],
+            "operator": "exact"
+        }],
+        "rollout_percentage": 20
+    }
+}
+```
+
+_At_ or _after_ the time specified, the change needs to be propagated onto the record.
```

Comment:
Indeed, there can be a backlog of tasks, and you need to be able to measure it and alert on it if the delay is causing too much degradation to the experience.
Easy to do by replicating the `ingestion_lag` celery task.